README
NAME
    Path::IsDev::HeuristicSet::FromConfig - A Custom Heuristic Set from a
    configuration file

VERSION
    version 0.001001

SYNOPSIS
        export PATH_ISDEV_DEFAULT_SET="FromConfig";
        # Whee, is_dev() and friends now seriously different everywhere

    By default, it will try to read a configuration file in one of the
    following paths:

        $HOME/.local/share/.path_isdev_heuristicset_fromconfig/config.json
        $HOME/.path_isdev_heuristicset_fromconfig/config.json

    Which ever one it is unfortunately dependent on the sort of mood
    File::HomeDir is in, and whether or not you have "xdg-user-dir" in your
    $PATH

    Either way, if such a path does not exist the first time you use this
    module, it will be created for you from the default template in the
    distributions share directory.

    Edit it to your liking.

    If you mess it up, just delete it, run the code again, and its back! :D

    In fact, its so aggressive at this, I had to put a bit of code in the
    tests to stop it creating those directories during tests >_>.

    Pester File::UserConfig if you want this logic improved.

METHODS
  "heuristics"
    Satisfies the role "HeuristicSet::Simple"

    Returns the values in the configuration file in the field "heuristics"

  "negative_heuristics"
    Satisfies the role "HeuristicSet::Simple"

    Returns the values in the configuration file in the field
    "negative_heuristics"

AUTHOR
    Kent Fredric <kentfredric@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Kent Fredric
    <kentfredric@gmail.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for Path-IsDev-HeuristicSet-FromConfig

{{$NEXT}}

0.002000 2013-10-18T11:17:13Z
 [00 Minor]
 - Include HomeDir and PerlINC in the default config.

 [Dependencies::Added / runtime requires]
 - Path::IsDev::NegativeHeuristic::HomeDir 1.000000
 - Path::IsDev::NegativeHeuristic::PerlINC 1.000000

 [Dependencies::Changed / develop requires]
 - Dist::Zilla::PluginBundle::Author::KENTNL v1.9.0 → 2.000000
 - Test::Kwalitee 1.08 → 1.12

 [Dependencies::Changed / test requires]
 - Test::More 0.98 → 0.99

0.001000 2013-10-06T22:33:06Z
 - First version.



lib/Path/IsDev/HeuristicSet/FromConfig.pm
use strict;
use warnings;

package Path::IsDev::HeuristicSet::FromConfig;
BEGIN {
  $Path::IsDev::HeuristicSet::FromConfig::AUTHORITY = 'cpan:KENTNL';
}
{
  $Path::IsDev::HeuristicSet::FromConfig::VERSION = '0.001001';
}

# ABSTRACT: A Custom Heuristic Set from a configuration file



use Role::Tiny::With;

with 'Path::IsDev::Role::HeuristicSet::Simple';

require Path::IsDev::HeuristicSet::FromConfig::Loader;

my $loader = Path::IsDev::HeuristicSet::FromConfig::Loader->new();


sub heuristics {
  return @{ $loader->heuristics() };
}


sub negative_heuristics {
  return @{ $loader->negative_heuristics() };
}

1;

__END__

=pod

=encoding utf-8

=head1 NAME

Path::IsDev::HeuristicSet::FromConfig - A Custom Heuristic Set from a configuration file

=head1 VERSION

version 0.001001

=head1 SYNOPSIS

    export PATH_ISDEV_DEFAULT_SET="FromConfig";
    # Whee, is_dev() and friends now seriously different everywhere

By default, it will try to read a configuration file in one of the following paths:

    $HOME/.local/share/.path_isdev_heuristicset_fromconfig/config.json
    $HOME/.path_isdev_heuristicset_fromconfig/config.json

Which ever one it is unfortunately dependent on the sort of mood File::HomeDir is in, and whether or not
you have C<xdg-user-dir> in your C<$PATH>

Either way, if such a path does not exist the first time you use this module, it will be created
for you from the default template in the distributions share directory.

Edit it to your liking.

If you mess it up, just delete it, run the code  again, and its back! :D

In fact, its so aggressive at this, I had to put a bit of code in the tests to stop it
creating those directories during tests >_>.

Pester File::UserConfig if you want this logic improved.

=head1 METHODS

=head2 C<heuristics>

Satisfies the role L<< C<HeuristicSet::Simple>|Path::IsDev::Role::HeuristicSet::Simple/heuristics >>

Returns the values in the configuration file in the field C<heuristics>

=head2 C<negative_heuristics>

Satisfies the role L<< C<HeuristicSet::Simple>|Path::IsDev::Role::HeuristicSet::Simple/negative_heuristics >>

Returns the values in the configuration file in the field C<negative_heuristics>

=begin MetaPOD::JSON v1.1.0

{
    "namespace":"Path::IsDev::HeuristicSet::FromConfig",
    "interface":"single_class",
    "does":"Path::IsDev::Role::HeuristicSet::Simple"
}


=end MetaPOD::JSON

=head1 AUTHOR

Kent Fredric <kentfredric@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Path/IsDev/HeuristicSet/FromConfig/Loader.pm

use strict;
use warnings;

package Path::IsDev::HeuristicSet::FromConfig::Loader;
BEGIN {
  $Path::IsDev::HeuristicSet::FromConfig::Loader::AUTHORITY = 'cpan:KENTNL';
}
{
  $Path::IsDev::HeuristicSet::FromConfig::Loader::VERSION = '0.001001';
}

# ABSTRACT: Configuration loader and decoder for C<::FromConfig>

sub _path {
  require Path::Tiny;
  goto \&Path::Tiny::path;
}


use Class::Tiny {
  dist             => sub { 'Path-IsDev-HeuristicSet-FromConfig' },
  module           => sub { 'Path::IsDev::HeuristicSet::FromConfig' },
  config_file      => sub { 'config.json' },
  config_file_full => sub { _path( $_[0]->config->configdir )->child( $_[0]->config_file ) },
  config           => sub {
    require File::UserConfig;
    return File::UserConfig->new(
      dist   => $_[0]->dist,
      module => $_[0]->module,
    );
  },
  decoder => sub {
    require JSON;
    return JSON->new();
  },
  data => sub {
    $_[0]->decoder->decode( $_[0]->config_file_full->slurp_utf8 );
  },
  heuristics => sub {
    return $_[0]->data->{heuristics} || [];
  },
  negative_heuristics => sub {
    return $_[0]->data->{negative_heuristics} || [];
  },
};








1;

__END__

=pod

=encoding utf-8

=head1 NAME

Path::IsDev::HeuristicSet::FromConfig::Loader - Configuration loader and decoder for C<::FromConfig>

=head1 VERSION

version 0.001001

=head1 ATTRIBUTES

=head2 C<dist>

The name of the C<dist> for C<sharedir> mechanics and C<config> paths.

    Path-IsDev-HeuristicSet-FromConfig

=head2 C<module>

The name of the C<module> for C<sharedir> mechanics and C<config> paths.

    Path::IsDev::HeuristicSet::FromConfig

=head2 C<config_file>

The name of the file relative to the configuration C<dir>

    config.json

=head2 C<config_file_full>

The full path to the C<config> file.

If not specified, combined from C<config> and C<config_file> wrapped in a C<Path::Tiny>

=head2 C<config>

Returns a C<File::UserConfig> object preconfigured with a few things ( namely, C<dist> and C<module> )

=head2 C<decoder>

Returns a C<JSON> object to perform decoding with

=head2 C<data>

Returns decoded data by slurping C<config_file_full> and throwing it in C<decoder>

=head2 C<heuristics>

Proxy for C<< data->{heuristics} >>

=head2 C<negative_heuristics>

Proxy for C<< data->{negative_heuristics} >>

=begin MetaPOD::JSON v1.1.0

{
    "namespace":"Path::IsDev::HeuristicSet::FromConfig::Loader",
    "interface":"class",
    "inherits":"Class::Tiny::Object"
}


=end MetaPOD::JSON

=head1 AUTHOR

Kent Fredric <kentfredric@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4007"
  },
  "configure_requires" => {
    "Module::Build" => "0.4007"
  },
  "dist_abstract" => "A Custom Heuristic Set from a configuration file",
  "dist_author" => [
    "Kent Fredric <kentfredric\@gmail.com>"
  ],
  "dist_name" => "Path-IsDev-HeuristicSet-FromConfig",
  "dist_version" => "0.001001",
  "license" => "perl",
  "module_name" => "Path::IsDev::HeuristicSet::FromConfig",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Class::Tiny" => 0,
    "File::UserConfig" => 0,
    "JSON" => 0,
    "Path::IsDev::NegativeHeuristic::HomeDir" => "1.000000",
    "Path::IsDev::NegativeHeuristic::PerlINC" => "1.000000",
    "Path::IsDev::Role::HeuristicSet::Simple" => 0,
    "Path::Tiny" => 0,
    "Role::Tiny::With" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "File::Spec" => 0,
    "FindBin" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Path::IsDev::Object" => 0,
    "Test::File::ShareDir" => 0,
    "Test::More" => "0.99"
  }
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
